### PR TITLE
Add source provenance to OpenMP parse tree nodes; clean things up.

### PR DESCRIPTION
**Pull request by klausler**
For Brian.  Ensure that OpenMP constructs all have source provenance (except ATOMIC, it needs more work).  Add `"..."_id` tokens that are safe from being misrecognized when appearing as prefixes to longer identifier-like things.  Remove needless usage of indirection in the parse tree nodes.  Convert unions of all empty classes into simple enums.  Restore alphabetic order where necessary.

### CODE REVIEW

File: lib/parser/dump-parse-tree.h [hunk 1]
```diff
@@ -488,93 +472,76 @@ class ParseTreeDumper {
...
+  NODE(parser, OmpScheduleModifier)
+  NODE(parser::OmpScheduleModifier, Modifier1)
+  NODE(parser::OmpScheduleModifier, Modifier2)
+  NODE(parser, OmpScheduleModifierType)
+  NODE_ENUM(parser::OmpScheduleModifierType, ModType)
   NODE(parser, OmpSection)
   NODE(parser, OmpStandaloneDirective)
-  NODE(parser, OpenMPCancellationPointConstruct)
+  NODE_ENUM(parser::OmpStandaloneDirective, Directive)
+  NODE(parser, Only)
```
Comment (ichoyjx) [inline lib/parser/dump-parse-tree.h]: This one is "mis-moved".

Comment (ichoyjx) [reply lib/parser/dump-parse-tree.h]: Sorry forget about it...

File: lib/parser/unparse.cc [hunk 2]
```diff
@@ -2060,29 +2060,25 @@ class UnparseVisitor {
...
-            [&](const OmpBlockDirective::TargetTeams &) {
-              Word("TARGET TEAMS ");
-            },
-            [&](const OmpBlockDirective::Target &) { Word("TARGET "); },
-            [&](const OmpBlockDirective::Taskgroup &) { Word("TASKGROUP "); },
-            [&](const OmpBlockDirective::Task &) { Word("TASK "); },
-            [&](const OmpBlockDirective::Teams &) { Word("TEAMS "); },
-        },
-        x.u);
+    switch (x.v) {
```
Comment (sscalpone) [inline lib/parser/unparse.cc]: What's the advantage of converting to a enum?

Comment (klausler) [reply lib/parser/unparse.cc]: Enum classes are simple, safe, get the job done, and you can use `switch` statements on them.  Using `std::variant` on a pile of empty class types violates the principle of least astonishment, forces the use `std::visit` / `std::holds_alternative` to decode, and adds no value.

File: lib/parser/openmp-grammar.h [hunk 3]
```diff
@@ -445,73 +416,69 @@ TYPE_PARSER("ATOMIC" >>
...
-// OMP TASKYIELD
-TYPE_PARSER("TASKYIELD" >> construct<OpenMPTaskyieldConstruct>() / endOmpLine)
-
 // OMP SINGLE
-TYPE_PARSER(skipStuffBeforeStatement >> "!$OMP "_sptok >> "END"_tok >>
-    construct<OmpEndSingle>("SINGLE"_tok >> Parser<OmpClauseList>{}))
+TYPE_PARSER(startOmpLine >> "END"_tok >>
+    construct<OmpEndSingle>("SINGLE" >> Parser<OmpClauseList>{}) / endOmpLine)
 
 TYPE_PARSER("SINGLE" >>
```
Comment (ichoyjx) [inline lib/parser/openmp-grammar.h]: `"SINGLE" >>` needs to be removed.

File: lib/parser/openmp-grammar.h [hunk 4]
```diff
@@ -445,73 +416,69 @@ TYPE_PARSER("ATOMIC" >>
...
+    construct<OpenMPSingleConstruct>(
+        OmpConstructDirective("SINGLE"_tok), block, Parser<OmpEndSingle>{}))
 
 // OMP WORKSHARE
-TYPE_PARSER("WORKSHARE" >>
-    construct<OpenMPWorkshareConstruct>(endOmpLine >> block,
-        Parser<OmpEndWorkshare>{} >>
-            maybe(construct<OmpNowait>("NOWAIT"_tok)) / endOmpLine))
+TYPE_PARSER(construct<OpenMPWorkshareConstruct>(
+    verbatim("WORKSHARE"_tok) / endOmpLine, block,
```
Comment (ichoyjx) [inline lib/parser/openmp-grammar.h]: It fails to get the source location using `verbatim`:
```
!$omp workshare
  a = 1.0
  !$omp end workshare
end
```
OR
```
b = 0.0
  !$omp sections
  !$omp section
  do i=1, 100
     do j=1, 200
        b = b + 1.0
     enddo
  enddo
  !$omp end sections nowait
end
```
Tree looks like:
`ExecutionPartConstruct -> ExecutableConstruct -> OpenMPConstruct -> OmpSection -> Verbatim = ''`

File: lib/parser/openmp-grammar.h [hunk 5]
```diff
@@ -520,76 +487,57 @@ TYPE_PARSER(construct<OmpEndDoSimd>(maybe(construct<OmpNowait>("NOWAIT"_tok))))
...
     construct<OmpEndSections>(
-        maybe("NOWAIT"_tok >> construct<OmpNowait>()) / endOmpLine))
+        maybe("NOWAIT" >> construct<OmpNowait>()) / endOmpLine))
 
 // OMP SECTIONS
-TYPE_PARSER("SECTIONS" >>
-    construct<OpenMPSectionsConstruct>(
-        Parser<OmpClauseList>{} / endOmpLine, block, Parser<OmpEndSections>{}))
+TYPE_PARSER(construct<OpenMPSectionsConstruct>(
+    OmpConstructDirective("SECTIONS"_tok), block, Parser<OmpEndSections>{}))
```
Comment (ichoyjx) [inline lib/parser/openmp-grammar.h]: `OmpConstructDirective("SECTIONS"_tok)` will save the source location for directive `SECTIONS` into `OmpClauseList`:
```
| | ExecutionPartConstruct -> ExecutableConstruct -> OpenMPConstruct -> OpenMPSectionsConstruct
| | | OmpClauseList = '**sections**'
| | | Block
| | | | ExecutionPartConstruct -> ExecutableConstruct -> OpenMPConstruct -> OmpSection -> Verbatim = ''
| | | | ExecutionPartConstruct -> ExecutableConstruct -> DoConstruct
```

**Comment by klausler** [comment]
Problems in review comments have been addressed.

**Comment by ichoyjx** [comment]
> Problems in review comments have been addressed.

Thanks! I am doing some tests and will let you know soon.

**Pull request merged**